design-proposals/klip-4-custom-types.md

```diff
@@ -0,0 +1,190 @@
+data with other parts of their organizations.
+
+## Scope
+
+* Syntax to register and unregister custom type aliases
+* Ability to include custom types at compile-time (e.g. in the JAR) for use in UDF declarations
+* Ability to list all custom types
+
+Schema Registry integration is not considered within scope.
+
+## Value/Return
+
+Real data is complex, nested and referential. This feature will help us dramatically improve the
+user interaction with complicated schemas. 
+
+## Public APIS
+
+### CREATE TYPE
+
+```sql
+CREATE TYPE <type_name> AS <type>;
+```
+
+The `CREATE TYPE` syntax will allow KSQL users to register a type alias directly in SQL (either 
+interactive or headless modes). Any types registered using this command can be leveraged in any
+future statement. 
+
+Any attempts to register the same type twice without a corresponding `DROP TYPE` will fail.
+
+### DROP TYPE 
+
+```sql
+DROP TYPE <type_name>;
+```
+
+The `DROP TYPE` syntax will allow KSQL users to remove a type alias from KSQL.
+
+### SHOW TYPES
+
+```sql
+SHOW TYPES;
+```
+
+The `SHOW TYPES` command will list all custom types and their type definitions. A sample output
+for this command would be:
+```
+|---------------|--------------------------------------------------------------|------------------|
+|     name      |       definition                                             | source           |
+|---------------|--------------------------------------------------------------|------------------|
+| ADDRESS       | STRUCT<number INTEGER, street VARCHAR, city VARCHAR>         | types.sql        |
+| PERSON        | STRUCT<firstname VARCHAR, lastname VARCHAR, address ADDRESS> | types.sql        |
+| COMPANY       | STRUCT<name VARCHAR, headquarters ADDRESS>                   | CLI              |
+|---------------|--------------------------------------------------------------|------------------|
+```
+
+### Extension Directory
+
+The `ksql.extension.dir` will now also recognize `.sql` files that contain only `CREATE TYPE`
+commands. These commands will all be run before compiling any UDFs so these custom type declarations
+can be used in the `@Udf` annotations. They will be loaded in natural order to ensure that the
+loading behavior is deterministic.
+
+## Design
+
+There will be a rewriter phase that will replace any aliased type with the full schema. For UDFs, 
+the annotation will lookup the schema directly and compile the UDF using the full schema. There
+are no restrictions on what types can be added as custom types, and custom types can be composite.
+
+This rewriting will be done _before_ enqueuing the command on the command topic to make sure that
+if the types change the existing statements will already be resolved. The user will see a success
+message that contains the rewritten schema.
+
+If a type is already registered via CLI, but is then added to the extensions directory, the value
```

Comment:
This is only possible if the server is restarted, right? I assume we only plan to load `.sql` files from the extensions directory when the KSQL server is started, as we currently do for UDFs.

Reply:
yes